File: compiler/IR.ml
open Ast

let _UPDATE = 0x10

(* Instruction Opcodes *)

let _NOP  = 0x00
let _CONS = 0x01
let _GLOB = 0x04
let _PRE  = 0x08
let _ADD  = 0x10
let _SUB  = 0x11
let _MUL  = 0x12
let _DIV  = 0x13
let _MOD  = 0x14
let _MIN  = 0x17
let _EQ   = 0x18
let _NEQ  = 0x19
let _GE   = 0x1A
let _LE   = 0x1B
let _GT   = 0x1C
let _LT   = 0x1D
let _AND =  0x20
let _OR   = 0x21
let _NOT  = 0x27
let _FBY  = 0x28
let _IF   = 0x40
let _CALL = 0x50
let _INDX = 0x54
let _FLD  = 0x58
let _POLY = 0x60
let _MONO = 0x68

(* Constant types *)

let _INT  = 0x01
let _BOOL = 0x02
let _GATE = 0x08
let _STR  = 0x10

let add_u8 buf u8 =
    Buffer.add_char buf @@ char_of_int u8

let add_u16 buf u16 =
    add_u8 buf ( u16        land 0xFF) ;
    add_u8 buf ((u16 lsr 8) land 0xFF)

let add_u32 buf u32 =
    add_u8 buf ( u32           land 0xFF) ;
    add_u8 buf ((u32 lsr 0x08) land 0xFF) ;
    add_u8 buf ((u32 lsr 0x10) land 0xFF) ;
    add_u8 buf ((u32 lsr 0x18) land 0xFF)

let add_i64 buf i64 =
    add_u8 buf ( i64           land 0xFF) ;
    add_u8 buf ((i64 asr 0x08) land 0xFF) ;
    add_u8 buf ((i64 asr 0x10) land 0xFF) ;
    add_u8 buf ((i64 asr 0x18) land 0xFF) ;
    add_u8 buf ((i64 asr 0x20) land 0xFF) ;
    add_u8 buf ((i64 asr 0x28) land 0xFF) ;
    add_u8 buf ((i64 asr 0x30) land 0xFF) ;
    add_u8 buf ((i64 asr 0x38) land 0xFF)

let add_pad buf byte count =
    for i = 1 to count do
        add_u8 buf byte
    done

let add_reloc relocs buf str =
    let seek = Buffer.length buf in
    add_u32 buf 0x00 ;
    relocs := (seek, str) :: !relocs

(* FIXME: completely inefficient.
 * should be pre-computed during toposort *)
let varnum v def =
    let equs = def.body.stmts in
    let v = match v with
        | Evar v -> v
        | _ -> assert false
    in let rec search equs = match equs with
        | [] ->
            (* FIXME: eww *)
            let rec search2 (inpts: signal list) = match inpts with
                | [] -> assert false
                | (hd :: tl) ->
                    if hd.name = v then 0
                    else 1 + search2 tl
            in search2 def.inputs
        | (Sif _ :: _) -> assert false
        | (Seq { lhs ; _ } :: tl) ->
            if lhs = v then 0
            else 1 + search tl
    in
    search equs

let add_align buf align =
    while not ((Buffer.length buf mod align) == 0) do
        add_u8 buf 0
    done

(* FIXME: we should not have to carry def along like that... *)
let add_simple_op buf opc args def =
    add_u8  buf opc ;
    add_u8  buf 0x0 ;
    List.iter (fun e -> add_u16 buf (varnum e def)) args ;
    add_align buf 8

let add_expr relocs buf e def = match e with
    | Ebinop (e1, Ofby, e2) -> add_simple_op buf _FBY [e1 ; e2] def
    | Ebinop (e1, Oadd, e2) -> add_simple_op buf _ADD [e1 ; e2] def
    | Ebinop (e1, Osub, e2) -> add_simple_op buf _SUB [e1 ; e2] def
    | Ebinop (e1, Omul, e2) -> add_simple_op buf _MUL [e1 ; e2] def
    | Ebinop (e1, Odiv, e2) -> add_simple_op buf _DIV [e1 ; e2] def
    | Ebinop (e1, Omod, e2) -> add_simple_op buf _MOD [e1 ; e2] def
    | Ebinop (e1, Oeq,  e2) -> add_simple_op buf _EQ  [e1 ; e2] def
    | Ebinop (e1, Oneq, e2) -> add_simple_op buf _NEQ [e1 ; e2] def
    | Ebinop (e1, Oge,  e2) -> add_simple_op buf _GE  [e1 ; e2] def
    | Ebinop (e1, Ole,  e2) -> add_simple_op buf _LE  [e1 ; e2] def
    | Ebinop (e1, Ogt,  e2) -> add_simple_op buf _GT  [e1 ; e2] def
    | Ebinop (e1, Olt,  e2) -> add_simple_op buf _LT  [e1 ; e2] def
    | Ebinop (e1, Oand, e2) -> add_simple_op buf _AND [e1 ; e2] def
    | Ebinop (e1, Oor,  e2) -> add_simple_op buf _OR  [e1 ; e2] def
    | Eunop (Ominus, e) -> add_simple_op buf _MIN [e] def
    | Eunop (Onot, e)   -> add_simple_op buf _NOT [e] def
    | Eif (c, t, f) -> add_simple_op buf _IF [c ; t ; f] def
    | Epre e -> add_simple_op buf _PRE [Evar e] def
    | Ecall (f, args) ->
        add_u8  buf _CALL ;
        add_u8  buf 0x00 ;
        add_u16 buf (List.length args) ;
        add_reloc relocs buf (Ident.show f) ;
        List.iter (fun e -> add_u16 buf (varnum e def)) args ;
        add_align buf 8
    | Evar v ->
        if List.mem_assoc v Normalize.glob_env then begin
            add_u8 buf _GLOB ;
            add_pad buf 0x0 3 ;
            add_reloc relocs buf (Ident.show v)
        end else add_simple_op buf _NOP [Evar v] def
    | Econst (Cnum i) ->
        add_u8 buf _CONS ;
        add_u8 buf _INT ;
        add_pad buf 0x0 6 ;
        add_i64 buf i
    | Econst (Cbool b) ->
        add_u8 buf _CONS ;
        add_u8 buf _BOOL ;
        add_u8 buf @@ if b then 0x01 else 0x00 ;
        add_align buf 8
    | Econst (Cstr s) ->
        add_u8 buf _CONS ;
        add_u8 buf _STR ;
        add_pad buf 0x0 2 ;
        add_reloc relocs buf s
    | Econst (Cgate g) ->
        add_u8 buf _CONS ;
        add_u8 buf _GATE ;
        add_u8 buf @@
        begin match g with
            | Goff -> 0x00
            | Gon  -> 0x01
            | Gtie -> 0xFF
        end ;                   (*  *)
        add_align buf 8 ;
    | Eindex (e1, e2) -> add_simple_op buf _INDX [e1 ; e2] def
    | Efield (e, f) ->
        add_u8 buf _FLD ;
        add_u8 buf 0x0 ;
        add_u16 buf (varnum e def) ;
        (* TODO: fields should be
         * compiled to int indices *)
        add_reloc relocs buf (Ident.show f)
    | Epoly s ->
        add_u8 buf _POLY ;
        add_u8 buf (List.length s) ;
        add_pad buf 0x0 6 ;
        List.iter
            (fun (e1, e2, e3) ->
                 add_u16 buf (varnum e1 def) ;
                 add_u16 buf (varnum e2 def) ;
                 add_u16 buf (varnum e3 def) ;
                 add_u16 buf 0x0)
            s
    | Emono (e1, e2, e3) ->
        add_u8 buf _MONO ;
        add_u8 buf 0x0 ;
        add_u16 buf (varnum e1 def) ;
        add_u16 buf (varnum e2 def) ;
        add_u16 buf (varnum e3 def)

let def definition =
    let msg = Buffer.create 512 in
    let relocs = ref [] in

    (* temporary msg size, will be filled later *)
    add_u32 msg 0 ;
    add_u32 msg _UPDATE ;
    add_reloc relocs msg (Ident.show definition.name) ;

    add_u8  msg (List.length definition.inputs) ;
    add_u8  msg (List.length definition.outputs) ;
    add_u16 msg (List.length definition.body.locals) ;

    (* var number of outputs, will be filled later *)
    (* 16 bits + 32 bits reserved *)
    add_pad msg 0x00 0x30 ;

    let equ_begin = Buffer.length msg in

    List.iter
        (function
            | Seq { rhs ; _ } -> add_expr relocs msg rhs definition
            | Sif _ -> assert false)
        definition.body.stmts ;

    let equ_size = Buffer.length msg - equ_begin in
    add_align msg 16 ;

    (* locals var nums (entry points) *)
    List.iter
        (fun v -> add_u16 msg (varnum (Evar v) definition))
        definition.body.locals ;

    add_align msg 16 ;
    let cst_begin = Buffer.length msg in

    (* constant data *)
    let relocs = List.fold_left
        (fun relocs (seek, str) ->
             let new_seek = Buffer.length msg in
             for i = 0 to String.length str - 1 do
                 Buffer.add_char msg str.[i]
             done ;
             Buffer.add_char msg (char_of_int 0) ;
             add_align msg 16 ;
             (seek, new_seek) :: relocs)
        [] !relocs
    in

    let cst_size = Buffer.length msg - cst_begin in

    (* freeze message, write relocations *)
    let size = Buffer.length msg in
    let msg = Buffer.to_bytes msg in

    let write_u16 addr u16 =
        Bytes.set msg  addr      @@ char_of_int @@  u16        land 0xFF ;
        Bytes.set msg (addr + 1) @@ char_of_int @@ (u16 lsr 8) land 0xFF ;
    in

    let write_u32 addr u32 =
        Bytes.set msg  addr      @@ char_of_int @@  u32           land 0xFF ;
        Bytes.set msg (addr + 1) @@ char_of_int @@ (u32 lsr 0x08) land 0xFF ;
        Bytes.set msg (addr + 2) @@ char_of_int @@ (u32 lsr 0x10) land 0xFF ;
        Bytes.set msg (addr + 3) @@ char_of_int @@ (u32 lsr 0x18) land 0xFF
    in

    (* message size *)
    write_u32 0x0 size ;

    (* outputs *)
    List.iteri
        (fun i (v: signal) ->
             write_u16 (0x10 + (i lsl 1)) (varnum (Evar v.name) definition))
        definition.outputs ;

    (* section sizes *)
    write_u32 0x20 equ_size ;
    write_u32 0x24 cst_size ;

    (* constant data addresses *)
    List.iter
        (fun (seek, new_seek) -> write_u32 seek new_seek)
        relocs ;
    msg

let command = function
    | Cdef d -> def d
    | _ -> failwith "unsupported message kind"

let to_ir = List.map command

File: new-compiler/src/ir.rs
use crate::ast;
use crate::util::BinWrite;
use crate::util::BinWriteExt;

const UPDATE: u32 = 0x10;

#[repr(u8)]
enum OpCode {
    Nop = 0x00,
    Cons = 0x01,
    Get = 0x02,
    Set = 0x03,
    // Glob = 0x04,
    Lambda = 0x05,
    // Pre = 0x08,
    Add = 0x10,
    Sub = 0x11,
    Mul = 0x12,
    Div = 0x13,
    Mod = 0x14,
    Min = 0x17,
    Eq = 0x18,
    Neq = 0x19,
    Ge = 0x1A,
    Le = 0x1B,
    Gt = 0x1C,
    Lt = 0x1D,
    And = 0x20,
    Or = 0x21,
    // Not = 0x27,
    // Fby = 0x28,
    If = 0x40,
    Call = 0x50,
    Indx = 0x54,
    Fld = 0x58,
    Poly = 0x60,
    Mono = 0x68
}

#[repr(u8)]
enum TypeCode {
    Int = 0x01,
    Bool = 0x02,
    Gate = 0x08,
    Str = 0x10,
}

struct TransCtx<'a> {
    buf: Vec<u8>,
    relocs: Vec<(usize, &'a str)>,
    next_node: u16,
    next_var: u16,
    depth: u16
}

impl BinWriteExt for Vec<u8> {
    fn len(&self) -> usize {
        self.len()
    }
}

use std::io::Write;

#[allow(unused_must_use)]
impl<'a> TransCtx<'a> {
    fn reloc(&mut self, data: &'a str) {
        let seek = self.buf.len();
        self.buf.write_le::<u32>(0);
        self.relocs.push((seek, data))
    }

    fn write_cst(&mut self, cst: &'a ast::Const) {
        self.buf.write_le::<u8>(OpCode::Cons as u8);
        match cst {
            ast::Const::Num(i) => {
                self.buf.write_le::<u8>(TypeCode::Int as u8);
                self.buf.pad(6);
                self.buf.write_le::<i64>(*i);
            }
            ast::Const::Bool(b) => {
                self.buf.write_le::<u8>(TypeCode::Bool as u8);
                self.buf.write_le::<u8>(if *b { 0x01 } else { 0x00 });
                // FIXME: could be pad 5?
                self.buf.align(8);
            }
            ast::Const::Str(ref s) => {
                self.buf.write_le::<u8>(TypeCode::Str as u8);
                self.buf.pad(2);
                self.reloc(&s);
            }
            ast::Const::Gate(g) => {
                self.buf.write_le::<u8>(TypeCode::Gate as u8);
                self.buf.write_le::<u8>(match g {
                    ast::Gate::Off => 0x00,
                    ast::Gate::On  => 0x01,
                    ast::Gate::Tie => 0x02,
                });
                // FIXME: could be pad 5?
                self.buf.align(8);
            }
        }
    }

    fn next(&mut self) -> u16 {
        let next = self.next_node;
        self.next_node += 1;
        next
    }

    fn simple_op(&mut self, op: OpCode, args: &[&'a ast::Expr]) -> u16 {
        let args = args.iter().map(|e| self.expr(e)).collect::<Vec<_>>();
        let this = self.next();
        self.buf.write_le::<u8>(op as u8);
        self.buf.write_le::<u8>(0);
        for a in args {
            self.buf.write_le::<u16>(a);
        }
        self.buf.align(8);
        this
    }

    fn expr(&mut self, e: &'a ast::Expr) -> u16 {
        use ast::ExprNode::*;
        use ast::BinOp::*;
        use ast::UnOp::*;
        match &*e.node {
            // simple operations (AST basic nodes)
            // binops
            BinOp(x, Add, y) => self.simple_op(OpCode::Add, &[x, y]),
            BinOp(x, Sub, y) => self.simple_op(OpCode::Sub, &[x, y]),
            BinOp(x, Mul, y) => self.simple_op(OpCode::Mul, &[x, y]),
            BinOp(x, Div, y) => self.simple_op(OpCode::Div, &[x, y]),
            BinOp(x, Mod, y) => self.simple_op(OpCode::Mod, &[x, y]),
            BinOp(x, Eq,  y) => self.simple_op(OpCode::Eq,  &[x, y]),
            BinOp(x, Neq, y) => self.simple_op(OpCode::Neq, &[x, y]),
            BinOp(x, Ge,  y) => self.simple_op(OpCode::Ge,  &[x, y]),
            BinOp(x, Le,  y) => self.simple_op(OpCode::Le,  &[x, y]),
            BinOp(x, Gt,  y) => self.simple_op(OpCode::Gt,  &[x, y]),
            BinOp(x, Lt,  y) => self.simple_op(OpCode::Lt,  &[x, y]),
            BinOp(x, And, y) => self.simple_op(OpCode::And, &[x, y]),
            BinOp(x, Or,  y) => self.simple_op(OpCode::Or,  &[x, y]),

            // unary ops
            UnOp(Plus, x)    => self.simple_op(OpCode::Nop, &[x]),
            UnOp(Minus, x)   => self.simple_op(OpCode::Min, &[x]),

            // basic constructs
            If(c, t, f)      => self.simple_op(OpCode::If,  &[c, t, f]),
            Index(e, i)      => self.simple_op(OpCode::Indx, &[e, i]),

            // special constructs (field access, poly and mono literals)

            Field(e, f) => {
                let e = self.expr(e);
                let this = self.next();
                self.buf.write_le::<u8>(OpCode::Fld as u8);
                self.buf.write_le::<u8>(0);
                self.buf.write_le::<u16>(e);
                // TODO: fields should be
                // compiled to int indices
                self.reloc(f.name());
                this
            }

            Poly(s) => {
                let vars = s.iter().map(|(e1, e2, e3)| {
                    let e1 = self.expr(e1);
                    let e2 = self.expr(e2);
                    let e3 = self.expr(e3);
                    (e1, e2, e3)
                }).collect::<Vec<_>>();
                let this = self.next();
                self.buf.write_le::<u8>(OpCode::Poly as u8);
                self.buf.write_le::<u8>(vars.len() as u8);
                self.buf.pad(6);
                for (e1, e2, e3) in vars {
                    self.buf.write_le::<u16>(e1);
                    self.buf.write_le::<u16>(e2);
                    self.buf.write_le::<u16>(e3);
                    self.buf.write_le::<u16>(0);
                }
                this
            }

            Mono(e1, e2, e3) => {
                let e1 = self.expr(e1);
                let e2 = self.expr(e2);
                let e3 = self.expr(e3);
                let this = self.next();
                self.buf.write_le::<u8>(OpCode::Mono as u8);
                self.buf.write_le::<u8>(0);
                self.buf.write_le::<u16>(e1);
                self.buf.write_le::<u16>(e2);
                self.buf.write_le::<u16>(e3);
                this
            }
            // vars, consts (AST leafs)
            Var(v) => {
                let this = self.next();
                self.buf.write_le(OpCode::Get);
                self.buf.write_le::<u8>(0);
                self.buf.write_le::<u16>(v.data().index.get());
                self.buf.write_le::<u16>(v.data().depth.get());
                self.buf.align(8);
                this
            }
            Cst(cst) => {
                let this = self.next();
                self.write_cst(cst);
                this
            }
            // AST nodes affecting env: let
            Let(v, e1, e2) => {
                let e1 = self.expr(e1);
                let nvar = self.next_var;
                self.next_var += 1;
                let e_set = self.next();
                self.buf.write_le::<u8>(OpCode::Set as u8);
                self.buf.write_le::<u8>(0);
                self.buf.write_le::<u16>(nvar);
                self.buf.write_le::<u16>(self.depth);
                self.buf.write_le::<u16>(e1);
                v.data().index.set(nvar);
                v.data().depth.set(self.depth);
                let e_let = self.expr(e2);
                // no need to restore anything since IDs are scoped
                // TODO: we could also restore var nums to reuse
                // memory spaces from different branches of AST
                let this = self.next();
                self.buf.write_le::<u8>(OpCode::Nop as u8);
                self.buf.write_le::<u8>(1);
                self.buf.write_le::<u16>(e_set);
                self.buf.write_le::<u16>(e_let);
                self.buf.align(8);
                this
            }
            Call(ef, e_args) => {
                let ef = self.expr(ef);
                let eargs = e_args.iter().map(|e| self.expr(e)).collect::<Vec<_>>();
                let this = self.next();
                self.buf.write_le::<u8>(OpCode::Call as u8);
                self.buf.write_le::<u8>(1);
                self.buf.write_le::<u16>(eargs.len() as u16);
                self.buf.write_le::<u16>(ef);
                self.buf.write_le::<u16>(0);
                for arg in eargs {
                    self.buf.write_le::<u16>(arg);
                }
                self.buf.align(8);
                this
            }
            Lambda(args, _, e) => {
                // push context frame
                self.depth += 1;
                let old_var = self.next_var;
                self.next_var = 0;
                for (v, _) in args {
                    v.data().index.set(self.next_var);
                    v.data().depth.set(self.depth);
                    self.next_var += 1;
                }
                let e = self.expr(e);
                let this = self.next();
                self.buf.write_le::<u8>(OpCode::Lambda as u8);
                self.buf.write_le::<u8>(args.len() as u8);
                self.buf.write_le::<u16>(self.next_var - args.len() as u16);
                self.buf.write_le::<u16>(e);
                self.buf.align(8);
                self.next_var = old_var;
                self.depth -= 1;
                this
            }
        }
    }

    fn new(buf: Vec<u8>) -> TransCtx<'a> {
        TransCtx { buf: buf, relocs: Vec::new(), next_var: 0,
                   next_node: 0, depth: 1 }
        // depth 0 is toplevel
    }
}

#[allow(unused_must_use)]
pub fn update(def: ast::FnDef) -> Vec<u8> {
    let buf = Vec::with_capacity(1024);
    let mut ctx = TransCtx::new(buf);

    // reserve space for msg size
    ctx.buf.write_le::<u32>(0);
    ctx.buf.write_le::<u32>(UPDATE);
    ctx.reloc(&def.name.name());

    ctx.buf.write_le::<u8>(def.inputs.len() as u8);
    ctx.buf.write_le::<u8>(def.outputs.len() as u8);
    // reloc
    ctx.buf.write_le::<u16>(0 as u16);

    // var number of outputs
    ctx.buf.pad(0x30);

    // attributes
    for ast::Attribute { name, args } in def.attrs.iter() {
        ctx.buf.write(name.name().as_bytes());
        ctx.buf.write_le(0u8);
        ctx.buf.write_le(args.len() as u8);
        ctx.buf.align(8);
        for arg in args {
            match arg {
                // FIXME
                // ast::PropertyArg::Id(id) => msg.write_cst(&ast::Const::Str(id)),
                ast::PropertyArg::Cst(cst) => {
                    ctx.write_cst(&cst);
                    // XXX: why align here and not above?
                    ctx.buf.align(8);
                }
                _ => ()
            }
        }
        ctx.buf.align(16);
    }

    match def.inputs {
        ast::Args::Named(ref args) => for (i, (v, _)) in args.iter().enumerate() {
            v.data().index.set(i as u16);
            v.data().depth.set(1);
            ctx.next_var += 1;
        }
        _ => panic!("function args should be named")
    }

    let instr_begin = ctx.buf.len() as u32;
    let e = ctx.expr(&def.body);
    let instr_size = ctx.buf.len() as u32 - instr_begin;
    ctx.buf.align(16);

    // locals var (none for now)
    ctx.next_var -= def.inputs.len() as u16;
    for _ in 0 .. ctx.next_var {
        ctx.buf.write_le::<u16>(0);
    }

    ctx.buf.align(16);
    let cst_begin = ctx.buf.len() as u32;

    let TransCtx { mut buf, relocs, .. } = ctx;

    // constant data
    let relocs = relocs.into_iter().map(|(seek, data)| {
        let new_seek = buf.len();
        buf.write(data.as_bytes());
        buf.write_le(0u8);
        buf.align(16);
        (seek, new_seek)
    }).collect::<Vec<_>>();

    let cst_size = buf.len() as u32 - cst_begin;
    let size = buf.len() as u32;

    // relocs
    // message size
    unsafe {
        use std::mem::transmute;
        *(transmute::<_, *mut u32>(buf.as_ptr())) = size;
        // locals: number of context variables
        *(transmute::<_, *mut u16>(buf[0x0E..].as_ptr())) = ctx.next_var;
        // outputs: we only have one and it is the function expr
        *(transmute::<_, *mut u16>(buf[0x10..].as_ptr())) = e;
        *(transmute::<_, *mut u32>(buf[0x20..].as_ptr())) = instr_size;
        *(transmute::<_, *mut u32>(buf[0x24..].as_ptr())) = cst_size;

        for (seek, new_seek) in relocs {
            *(transmute::<_, *mut u32>(buf[seek as usize..].as_ptr())) =
                new_seek as u32;
        }
    }

    buf
}
